src/Test/Utilities/Shared/Compilation/IRuntimeEnvironment.cs

```diff
@@ -0,0 +1,334 @@
+
+            foreach (var referencedCompilation in referencedCompilations)
+            {
+                var emitData = RuntimeUtilities.EmitCompilationCore(referencedCompilation, null, diagnostics, null);
+                if (emitData.HasValue)
+                {
+                    var moduleData = new ModuleData(referencedCompilation.Assembly.Identity,
+                                                    OutputKind.DynamicallyLinkedLibrary,
+                                                    emitData.Value.Assembly,
+                                                    pdb: default(ImmutableArray<byte>),
+                                                    inMemoryModule: true);
+                    fullNameSet.Add(moduleData.Id.FullName);
+                    dependencies.Add(moduleData);
+                }
+            }
+
+            // Now that the Compilation values have been emitted, emit the non-compilation references
+            foreach (var current in (new[] { compilation }).Concat(referencedCompilations))
+            {
+                EmitReferences(current, fullNameSet, dependencies, diagnostics);
+            }
+
+            return RuntimeUtilities.EmitCompilationCore(compilation, manifestResources, diagnostics, testData);
+        }
+
+        internal static EmitOutput? EmitCompilationCore(
+            Compilation compilation,
+            IEnumerable<ResourceDescription> manifestResources,
+            DiagnosticBag diagnostics,
+            CompilationTestData testData
+        )
+        {
+            using (var executableStream = new MemoryStream())
+            {
+                var pdb = default(ImmutableArray<byte>);
+                var assembly = default(ImmutableArray<byte>);
+                var pdbStream = MonoHelpers.IsRunningOnMono()
+                    ? null
+                    : new MemoryStream();
+
+                EmitResult result;
+                try
+                {
+                    result = compilation.Emit(
+                        executableStream,
+                        pdbStream: pdbStream,
+                        xmlDocumentationStream: null,
+                        win32Resources: null,
+                        manifestResources: manifestResources,
+                        options: EmitOptions.Default,
+                        debugEntryPoint: null,
+                        testData: testData,
+                        getHostDiagnostics: null,
+                        cancellationToken: default(CancellationToken));
+                }
+                finally
+                {
+                    if (pdbStream != null)
+                    {
+                        pdb = pdbStream.ToImmutable();
+                        pdbStream.Dispose();
+                    }
+                }
+
+                diagnostics.AddRange(result.Diagnostics);
+                assembly = executableStream.ToImmutable();
+
+                if (result.Success)
```

Comment:
Seems a little strange that we spend time building up `pdb` and `assembly` as immutable arrays, then possibly throw them away if the compilation failed.